execution-core/src/stake.rs

```diff
@@ -24,107 +25,140 @@ pub const STAKE_WARNINGS: u8 = 1;
 
 /// Calculate the block height at which the next epoch takes effect.
 #[must_use]
-pub const fn next_epoch(block_height: BlockHeight) -> u64 {
+pub const fn next_epoch(block_height: u64) -> u64 {
     let to_next_epoch = EPOCH - (block_height % EPOCH);
     block_height + to_next_epoch
 }
 
 /// Stake a value on the stake contract.
 #[derive(Debug, Clone, PartialEq, Eq, Archive, Serialize, Deserialize)]
-#[archive_attr(derive(bytecheck::CheckBytes))]
+#[archive_attr(derive(CheckBytes))]
 pub struct Stake {
-    /// Public key to which the stake will belong.
-    pub public_key: StakePublicKey,
-    /// Signature belonging to the given public key.
-    pub signature: StakeSignature,
-    /// Value to stake.
-    pub value: u64,
+    payload: StakePayload,
+    signature: AccountSignature,
 }
 
 impl Stake {
-    const MESSAGE_SIZE: usize = u64::SIZE + u64::SIZE;
-    /// Return the digest to be signed in the `stake` function of the stake
-    /// contract.
+    /// Create a new stake.
     #[must_use]
-    pub fn signature_message(
-        counter: u64,
-        value: u64,
-    ) -> [u8; Self::MESSAGE_SIZE] {
-        let mut bytes = [0u8; Self::MESSAGE_SIZE];
+    pub fn new(payload: StakePayload, signature: AccountSignature) -> Self {
+        Self { payload, signature }
+    }
 
-        bytes[..u64::SIZE].copy_from_slice(&counter.to_bytes());
-        bytes[u64::SIZE..].copy_from_slice(&value.to_bytes());
+    /// Payload of the stake.
+    #[must_use]
+    pub fn payload(&self) -> &StakePayload {
+        &self.payload
+    }
 
-        bytes
+    /// Signature of the stake.
+    #[must_use]
+    pub fn signature(&self) -> &AccountSignature {
+        &self.signature
+    }
+
+    /// Return the message that should be used as the input to the signature.
+    ///
+    /// Just a proxy to [`StakePayload::signature_message`].
+    #[must_use]
+    pub fn signature_message(&self) -> [u8; StakePayload::MESSAGE_SIZE] {
+        self.payload.signature_message()
     }
 }
 
-/// Unstake a value from the stake contract.
-#[derive(Debug, Clone, PartialEq, Archive, Deserialize, Serialize)]
+/// The payload of a [`Stake`].
```

Comment:
These are not transactions - only Moonlight and Phoenix transactions should be referred to as such. I use the word "call" instead, since it's more accurate,